### PR TITLE
HAPI Integration Refactor

### PR DESCRIPTION
**Pull request by nnance**
This is a significant refactor of the HAPI integration to make the plugins more flexible and idiomatic.  This should resolve the outstanding issues related to the HAPI integration.  Specifically ( #97 and #123 )

WARNING: This refactor does result in breaking changes to the API.

TODO:

- [x] Update CHANGELOG.md with your change (include reference to issue & this PR)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass

### CODE REVIEW

**Comment by nnance** [comment]
@helfer The reason for a drop in code coverage is because I resolved one of the todo items which is to handle JSON parse errors in variables.  As a result there isn't a test for that case.  I have a set of changes that includes a test and fixes in all other integrations that increases the code coverage.  Do you want me to include the changes in the PR or create a new one?

**Comment by dahjelle** [comment]
Does `options` still accept a Promise?

**Comment by nnance** [comment]
@dahjelle no it doesn't.  I just realized that my example in the readme is wrong.  If you look at the integration test you can get an idea of how it is used but now options include { path, route, apolloOptions }  where apolloOptions can still be a promise.  I will fix the example now.

**Comment by dahjelle** [comment]
I can confirm that it appears to address the issue in #123. Hooray! :-D

File: src/integrations/hapiApollo.ts [hunk 1]
```diff
@@ -137,9 +152,75 @@ async function processQuery(body, optionsObject) {
...
+  return reply(responses);
 }
 
 function isOptionsFunction(arg: ApolloOptions | HAPIOptionsFunction): arg is HAPIOptionsFunction {
   return typeof arg === 'function';
 }
+
+function createErr(code: number, message: string) {
+  const err = Boom.create(code);
+  err.output.payload.message = message;
```
Comment (SimonDegraeve) [inline src/integrations/hapiApollo.ts]: Why do we need this? The `err.output.payload.message` is  derived from `err.message`

**Comment by SimonDegraeve** [comment]
I like the use of the route prerequisites, it looks really clean.

Still I have a question, why using `HAPI` (uppercase) everywhere? It is not an acronym. On the project's website they always use lowercase.

File: src/integrations/hapiApollo.ts (hunk 1 above)
Comment (nnance) [inline src/integrations/hapiApollo.ts]: so I struggled with this for a while.  When I use Boom.create(500, 'Something went wrong') the payload returned simply says `Internal server error`.  I had to do this to change the message of the payload.  Thoughts?

**Comment by nnance** [comment]
@SimonDegraeve as I was using the all uppercase version I wondered which one was right.  I agree I will lowercase hapi in my next commit.

File: src/integrations/hapiApollo.ts (hunk 1 above)
Comment (SimonDegraeve) [inline src/integrations/hapiApollo.ts]: I just double check and you are right. For an error 500, the message is overwritten.
See [https://github.com/hapijs/boom/blob/master/lib/index.js#L136-L138](https://github.com/hapijs/boom/blob/master/lib/index.js#L136-L138)

**Comment by helfer** [comment]
This is awesome! As far as I'm concerned, we can merge this as soon as we have a corresponding PR on the docs, which shouldn't take more than 10 minutes of work.

:tada:

**Comment by nnance** [comment]
@SimonDegraeve @helfer I have decided to handle the renaming of Hapi #129 and code coverage in separate pull requests.  There are a couple of other changes that are tagged for the 0.3 release that I will be working on today as well.  I will update the docs once all the 0.3 are complete.

**Pull request merged**